src/jdk.compiler/share/classes/com/sun/tools/javac/parser/UnicodeReader.java

```diff
@@ -25,267 +25,500 @@
+        width = 0;
+
+        // Fetch next character.
+        nextCodeUnit();
+
+        // If second backslash is detected.
+        if (wasBackslash) {
+            // Treat like a normal character (not part of unicode escape.)
+            wasBackslash = false;
+        } else if (character == '\\') {
+            // May be an unicode escape.
+            wasBackslash = !unicodeEscape();
         }
+
+        // Codepoint and character match if not surrogate.
+        codepoint = (int)character;
     }
 
-    /** Read next character in comment, skipping over double '\' characters.
+    /**
+     * Fetches the nextcode point from the buffer. If an unicode escape is recognized
+     * then converts unicode escape to a character. If two characters are a surrogate pair
+     * then converts to a codepoint.
      */
-    protected void scanCommentChar() {
-        scanChar();
-        if (ch == '\\') {
-            if (peekChar() == '\\' && !isUnicode()) {
-                skipChar();
-            } else {
-                convertUnicode();
+    private void nextCodePoint() {
+        // Next unicode character.
+        nextUnicodeInputCharacter();
+
+        // Return early if ASCII or not a surrogate pair.
+        if (isASCII() || !Character.isHighSurrogate(character)) {
+            return;
+        }
+
+        // Capture high surrogate and position.
+        char hi = character;
+        int savePosition = position;
+        int saveWidth = width;
+
+        // Get potential low surrogate.
+        nextUnicodeInputCharacter();
+        char lo = character;
+
+        if (Character.isLowSurrogate(lo)) {
+            // Start codepoint at start of high surrogate.
+            position = savePosition;
+            width += saveWidth;
+            // Compute codepoint.
+            codepoint = Character.toCodePoint(hi, lo);
+        } else {
+            // Restore to treat high surrogate as just a character.
+            position = savePosition;
+            width = saveWidth;
+            character = hi;
+            codepoint = (int)hi;
+            // Could potential report an error here (old code did not.)
+        }
+    }
+
+    /**
+     * Converts an unicode escape into a character.
+     *
+     * @return true if was an unicode escape.
+     */
+    private boolean unicodeEscape() {
+        // Start of unicode escape (past backslash.)
+        int start = position + width;
+
+        // Default to backslash result, unless proven otherwise.
+        character = '\\';
+        width = 1;
+
+        // Skip multiple 'u'.
+        int index;
+        for (index = start; index < length; index++) {
+            if (buffer[index] != 'u') {
+                break;
             }
         }
+
+        // Needs to have been at least one u.
+        if (index == start) {
+            return false;
+        }
+
+        int code = 0;
+
+        for (int i = 0; i < 4; i++) {
+            int digit = Character.digit(buffer[index], 16);
```

Comment:
This looks suspicious - what if index ends up being bigger than (or equal to)  `buffer.length` ?
Maybe we need a test for incomplete unicode sequences at the end of the tokenizer input - e.g. `\u123`

Reply:
You are correct. Will revise.

Reply:
Added change to 

    int digit = index < length ? Character.digit(buffer[index], 16) : -1;

Also added new test.